Mind failing linting issues:
```
Error: netboxlabs/diode/sdk/chunking.py:3:1: D213 Multi-line docstring summary should start at the second line
Error: netboxlabs/diode/sdk/chunking.py:10:1: UP035 Import from `collections.abc` instead: `Iterable`
Error: netboxlabs/diode/sdk/chunking.py:18:5: D213 Multi-line docstring summary should start at the second line
Error: netboxlabs/diode/sdk/chunking.py:35:5: D413 Missing blank line after last section ("Examples")
Error: netboxlabs/diode/sdk/chunking.py:88:5: D213 Multi-line docstring summary should start at the second line
Error: netboxlabs/diode/sdk/chunking.py:99:5: D413 Missing blank line after last section ("Examples")
Error: tests/test_chunking.py:182:5: D213 Multi-line docstring summary should start at the second line
```

--- a/README.md
+++ b/README.md
@@ -281,6 +281,78 @@ client = DiodeClient(
 )
 ```
 
+### Message chunking
+
+When ingesting large numbers of entities, you may need to split them into smaller chunks to avoid exceeding the gRPC message size limit for a single `ingest()` call. The SDK provides chunking utilities that automatically split entity lists into appropriately sized chunks.
+
+#### How it works
+
+The SDK uses a **greedy bin-packing algorithm** that:
+1. Accumulates entities until adding the next entity would exceed the size limit
+2. Starts a new chunk when the limit would be exceeded
+3. Ensures each chunk stays safely under the configured limit (default: 3 MB)
+
+#### Basic usage
+
+```python
+from netboxlabs.diode.sdk import DiodeClient, create_message_chunks
+from netboxlabs.diode.sdk.ingester import Device, Entity
+
+with DiodeClient(
+    target="grpc://localhost:8080/diode",
+    app_name="my-app",
+    app_version="1.0.0",
+) as client:
+    # Create a large list of entities
+    entities = []
+    for i in range(10000):
+        device = Device(
+            name=f"Device {i}",
+            device_type="Device Type A",
+            site="Site ABC",
+            role="Role ABC",
+        )
+        entities.append(Entity(device=device))
+
+    # Split into chunks (default 3 MB per chunk), then ingest each chunk separately.
+    for chunk in create_message_chunks(entities):
+        client.ingest(entities=chunk)
+```
+
+#### Custom chunk size
+
+You can customize the chunk size if needed:
+
+```python
+from netboxlabs.diode.sdk import create_message_chunks
+
+# Use a larger chunk size (3.5 MB)
+chunks = create_message_chunks(entities, max_chunk_size_mb=3.5)
+
+# Use a smaller chunk size for conservative chunking (2 MB)
+chunks = create_message_chunks(entities, max_chunk_size_mb=2.0)
+```
+
+#### Estimating message size
+
+You can estimate the serialized size of entities before chunking:
+
+```python
+from netboxlabs.diode.sdk import estimate_message_size
+
+size_bytes = estimate_message_size(entities)
+size_mb = size_bytes / (1024 * 1024)
+print(f"Total size: {size_mb:.2f} MB")
+
+# Decide whether chunking is needed
+if size_mb > 3.0:
+    chunks = create_message_chunks(entities)
+else:
+    # Small enough to send in one request
+    client.ingest(entities=entities)
+```
+
+
 ### Dry run mode
 
 `DiodeDryRunClient` generates ingestion requests without contacting a Diode server. Requests are printed to stdout by default, or written to JSON files when `output_dir` (or the `DIODE_DRY_RUN_OUTPUT_DIR` environment variable) is specified. The `app_name` parameter serves as the filename prefix; if not provided, `dryrun` is used as the default prefix. The file name is suffixed with a nanosecond-precision timestamp, resulting in the format `<app_name>_<timestamp_ns>.json`.

--- a/README.md
+++ b/README.md
@@ -346,7 +346,8 @@ print(f"Total size: {size_mb:.2f} MB")
 
 # Decide whether chunking is needed
 if size_mb > 3.0:
-    chunks = create_message_chunks(entities)
+    for chunk in create_message_chunks(entities):
+        client.ingest(entities=chunk)
 else:
     # Small enough to send in one request
     client.ingest(entities=entities)

--- a/netboxlabs/diode/sdk/chunking.py
+++ b/netboxlabs/diode/sdk/chunking.py
@@ -1,13 +1,14 @@
 #!/usr/bin/env python
 # Copyright 2026 NetBox Labs Inc
-"""Message chunking utilities for Diode SDK.
+"""
+Message chunking utilities for Diode SDK.
 
 This module provides utilities for chunking large lists of entities into
 size-appropriate chunks for gRPC ingestion, ensuring no chunk exceeds
 the gRPC message size limit.
 """
 
-from typing import Iterable
+from collections.abc import Iterable
 
 from .diode.v1 import ingester_pb2
 
@@ -15,7 +16,8 @@ from .diode.v1 import ingester_pb2
 def create_message_chunks(
     entities: Iterable[ingester_pb2.Entity], max_chunk_size_mb: float = 3.0
 ) -> list[list[ingester_pb2.Entity]]:
-    """Create size-aware chunks from entities using greedy bin-packing.
+    """
+    Create size-aware chunks from entities using greedy bin-packing.
 
     This function chunks entities to ensure each chunk stays under the specified
     size limit. It uses a greedy bin-packing algorithm that accumulates entities
@@ -40,6 +42,7 @@ def create_message_chunks(
 
         >>> # Use a custom chunk size
         >>> chunks = create_message_chunks(entities, max_chunk_size_mb=3.5)
+
     """
     # Convert iterable to list if necessary for size estimation
     if not isinstance(entities, list):
@@ -85,7 +88,8 @@ def create_message_chunks(
 
 
 def estimate_message_size(entities: Iterable[ingester_pb2.Entity]) -> int:
-    """Estimate the serialized size of entities in bytes.
+    """
+    Estimate the serialized size of entities in bytes.
 
     Calculates the total size by summing individual entity sizes plus the
     IngestRequest protobuf overhead.
@@ -101,6 +105,7 @@ def estimate_message_size(entities: Iterable[ingester_pb2.Entity]) -> int:
         >>> size_bytes = estimate_message_size(entities)
         >>> size_mb = size_bytes / (1024 * 1024)
         >>> print(f"Estimated size: {size_mb:.2f} MB")
+
     """
     # Convert iterable to list if necessary
     if not isinstance(entities, list):

--- a/netboxlabs/diode/sdk/chunking.py
+++ b/netboxlabs/diode/sdk/chunking.py
@@ -1,5 +1,5 @@
 #!/usr/bin/env python
-# Copyright 2024 NetBox Labs Inc
+# Copyright 2026 NetBox Labs Inc
 """Message chunking utilities for Diode SDK.
 
 This module provides utilities for chunking large lists of entities into

--- a/netboxlabs/diode/sdk/client.py
+++ b/netboxlabs/diode/sdk/client.py
@@ -668,7 +668,7 @@ class DiodeOTLPClient(DiodeClientInterface):
                 options=channel_opts,
             )
         else:
-            _LOGGER.debug(f"Setting up gRPC insecure channel")
+            _LOGGER.debug("Setting up gRPC insecure channel")
             base_channel = grpc.insecure_channel(
                 target=self._target,
                 options=channel_opts,

--- a/tests/test_chunking.py
+++ b/tests/test_chunking.py
@@ -1,5 +1,5 @@
 #!/usr/bin/env python
-# Copyright 2024 NetBox Labs Inc
+# Copyright 2026 NetBox Labs Inc
 """Tests for message chunking utilities."""
 
 from unittest.mock import patch

--- a/tests/test_chunking.py
+++ b/tests/test_chunking.py
@@ -179,7 +179,8 @@ def test_create_message_chunks_with_iterable():
 
 
 def test_create_message_chunks_single_large_entity():
-    """Test create_message_chunks with a single entity that exceeds chunk size.
+    """
+    Test create_message_chunks with a single entity that exceeds chunk size.
 
     This edge case verifies the function doesn't fail when a single entity
     is larger than the chunk size limit.

